src/main/kotlin/com/swisscom/health/des/cdr/clientvm/scheduling/DocumentUploadSchedulers.kt

```diff
@@ -0,0 +1,293 @@
+import kotlinx.coroutines.flow.onCompletion
+import kotlinx.coroutines.flow.onEach
+import kotlinx.coroutines.flow.shareIn
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.time.delay
+import kotlinx.coroutines.withContext
+import org.springframework.beans.factory.annotation.Qualifier
+import org.springframework.beans.factory.annotation.Value
+import org.springframework.context.annotation.Profile
+import org.springframework.scheduling.annotation.Scheduled
+import org.springframework.stereotype.Service
+import java.nio.file.Files
+import java.nio.file.Path
+import java.nio.file.attribute.BasicFileAttributes
+import java.util.Objects.isNull
+import java.util.concurrent.TimeUnit
+import kotlin.coroutines.coroutineContext
+import kotlin.io.path.absolute
+import kotlin.io.path.absolutePathString
+import kotlin.io.path.extension
+import kotlin.io.path.isRegularFile
+import kotlin.io.path.listDirectoryEntries
+import kotlin.io.path.name
+
+private val logger = KotlinLogging.logger {}
+
+@Service
+@Profile("!noEventTriggerUploadScheduler")
+class EventTriggerUploadScheduler(
+    private val config: CdrClientConfig,
+    private val tracer: Tracer,
+    @Qualifier("limitedParallelismCdrUploadsDispatcher")
+    private val cdrUploadsDispatcher: CoroutineDispatcher,
+    @Value("\${management.tracing.sampling.probability:1.0}")
+    private val samplerProbability: Double,
+    pushFileHandling: PushFileHandling,
+    processingInProgressCache: ObjectKache<String, Path>,
+) : BaseUploadScheduler(
+    config = config,
+    pushFileHandling = pushFileHandling,
+    cdrUploadsDispatcher = cdrUploadsDispatcher,
+    processingInProgressCache = processingInProgressCache,
+    tracer = tracer,
+) {
+
+    @PostConstruct
+    @Suppress("UnusedPrivateMember")
+    private fun failIfTelemetrySamplingIsEnabled() {
+        if (samplerProbability > ZERO_SAMPLING_THRESHOLD) {
+            logger.error {
+                "Telemetry sampling is enabled (sampling probability is set to $samplerProbability). Currently we cannot support telemetry " +
+                        "sampling without introducing a memory leak due to the lack of framework integration of micrometer/open-telemetry with Kotlin " +
+                        "coroutines/asynchronous flows. You need to disable telemetry sampling."
+            }
+            error("Telemetry sampling is enabled. Please set the configuration property `management.tracing.sampling.probability` to 0.0")
+        }
+    }
+
+    // NOTE: The short initial delay is there to make it as likely as possible for the task to be already running by the time the SpringBoot test
+    // sets the per-test configuration on the spy of the `CdrClientConfig`, which includes the test source directory. If the original task starts
+    // too late it will pick up the test source directory, and we have two instances of the watcher consuming events from that directory.
+    @Scheduled(initialDelay = 0L, timeUnit = TimeUnit.MILLISECONDS)
```

Comment:
The default for initialDelay is "-1". I haven't found any information what would be the difference when it is set to "0"? Does that really have any impact?

Reply:
It's because `org.springframework.scheduling.annotation.ScheduledAnnotationBeanPostProcessor` has a check that does not allow for negative initial delays (even though the scheduler, on which the task is then scheduled, is perfectly fine with negative delays).